### PR TITLE
Transition from jgit to git

### PR DESCRIPTION
**Pull request by val-fed**

### CODE REVIEW

File: common/repo/repo.proto [hunk 1]
```diff
@@ -29,6 +29,7 @@ message File {
     RENAME = 2;
     MODIFY = 3;
     COPY = 4;
+    UNTRACKED = 5;
```
Comment (oferb) [inline common/repo/repo.proto]: Let's not add UNTRACKED, and instead use the other ones (e.g ADD).
I understand that ADD can be tracked or untracked, but that distinction is not important at this point.

File: common/repo/GitRepo.java [hunk 2]
```diff
@@ -98,7 +67,26 @@ private CommandResult runCommand(String command) {
     CommandResult result = new CommandResult();
     try {
       List<String> fullCommand = new ArrayList<>(gitCommandBase);
-      fullCommand.addAll(Arrays.asList(command.split(" ")));
+      if (!command.contains("\"")) {
```
Comment (oferb) [inline common/repo/GitRepo.java]: Why did you need all of this new code?

File: common/repo/GitRepo.java [hunk 3]
```diff
@@ -169,42 +156,25 @@ public void tagHead(String name) {
...
-    } catch (GitAPIException e) {
-      throw new RuntimeException(e);
+    CommandResult commandResult = runCommand("status --short");
+    ImmutableList<String> lines = splitLines(commandResult.stdout);
+    for (String line : lines) {
+      String[] parts;
+      if (!line.trim().startsWith("M")
+          || !line.trim().startsWith("RM")
+          || !line.trim().startsWith("??")) {
+        line = line.replaceFirst("\\s{2}", " ");
```
Comment (oferb) [inline common/repo/GitRepo.java]: Can you put `"\\s{2}"` in a variable and name that variable to explain what the regex does?

File: common/repo/GitRepo.java [hunk 4]
```diff
@@ -248,107 +225,73 @@ public void tagHead(String name) {
...
-      // reset to `ref` and all changes introduced after it
-      // would be marked as unstaged but saved in working tree
-      jGit.reset().setMode(ResetType.MIXED).setRef(ref).call();
-    } catch (GitAPIException e) {
-      throw new RuntimeException(e);
-    }
+    // HEAD pointer would be
+    // reset to `ref` and all changes introduced after it
+    // would be marked as unstaged but saved in working tree
+    runCommand("reset " + ref);
```
Comment (oferb) [inline common/repo/GitRepo.java]: We don't need it mixed anymore?
@vmax WDYT?

File: common/repo/GitRepo.java [hunk 5]
```diff
@@ -248,107 +225,73 @@ public void tagHead(String name) {
...
+    CommandResult commandResult = runCommand("merge " + branch + " -s recursive -Xtheirs");
+    String err = commandResult.stderr;
+    return err.length() == 0;
+  }
+
+  public boolean merge(String branch, boolean remote) {
+    switchToMasterBranch();
+    CommandResult commandResult;
+    if (remote) {
+      runCommand("fetch origin " + branch);
```
Comment (oferb) [inline common/repo/GitRepo.java]: What if `fetch origin` fails?

File: common/repo/GitRepo.java (hunk 2 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: It needs for `commit` command.
This command in our case has next syntax:
`commit -m "commit message"`
If we will split the command just using one space delimiter( command.split(" ") ) we get next parts:
[0] commit
[1] -m
[2] "commit
[3] message"

This code checks the command. If the command contains `"` character it groups all words between `"...."` characters in one part:
[0] commit
[1] -m
[2] commit message

File: common/repo/GitRepo.java (hunk 5 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: I think we can throw `IllegasStateException` if `fetch origin` fails.
```
      CommandResult fetchCommandResult = runCommand("fetch origin " + branch);
      if (fetchCommandResult.stderr.length() != 0) {
        throw new IllegalStateException(
            "Failed to merge remote branch \'" + branch + "\': " + fetchCommandResult.stderr);
      }
```
What do you think about it?

File: common/repo/GitRepo.java (hunk 4 above)
Comment (vmax) [reply common/repo/GitRepo.java]: @oferb `mixed` is the default mode (*). we could state it explicitly if you so wish but it won't change the logic

>(*) If you specify the --mixed option, reset will stop at this point. This is also the default, so if you specify no option at all (just git reset HEAD~ in this case), this is where the command will stop.
>
>from https://git-scm.com/book/en/v2/Git-Tools-Reset-Demystified

Comment (oferb) [reply common/repo/GitRepo.java]: Ok, great :)
No need to state it then.

File: common/repo/GitRepo.java (hunk 2 above)
Comment (oferb) [reply common/repo/GitRepo.java]: We start off with `"commit message"` in a separate string, then we combine it, then we work hard to split it again.

Instead, let's have another method `runCommand()` that gets a `List<String> command` that's already split, and for `commit` command, call it?
The current `runCommand(String command)` would trivially split based on spaces and call `runCommand(List<String> command)`

File: common/repo/GitRepo.java (hunk 5 above)
Comment (oferb) [reply common/repo/GitRepo.java]: Sounds good

File: common/repo/GitRepo.java [hunk 6]
```diff
@@ -16,65 +16,36 @@
...
-import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
-import org.eclipse.jgit.treewalk.TreeWalk;
 
 // TODO: Implement methods
 @AutoFactory
 public class GitRepo implements Repo {
-  private final Git jGit;
-  private final Repository jGitRepo;
+  private static final String TWO_WHITE_SPACES = "\\s{2}";
+  private static final String DOUBLE_QUOTE = "\"";
```
Comment (oferb) [inline common/repo/GitRepo.java]: It's just a simple string and used only once, so I wouldn't make a `static final` from it.

File: common/repo/GitRepo.java [hunk 7]
```diff
@@ -150,11 +122,12 @@ public void tagHead(String name) {
   }
 
   public ImmutableList<String> getCommitIds(String branch) {
-    CommandResult commandResult = runCommand("log --pretty=%H master~1.." + branch);
+    CommandResult commandResult = runCommand("log --pretty=%H " + branch);
```
Comment (oferb) [inline common/repo/GitRepo.java]: You don't need `master..`?

File: common/repo/GitRepo.java [hunk 8]
```diff
@@ -169,42 +142,25 @@ public void tagHead(String name) {
...
-              untracked ->
-                  files.add(
-                      File.newBuilder().setAction(File.Action.ADD).setFilename(untracked).build()));
-    } catch (GitAPIException e) {
-      throw new RuntimeException(e);
+    CommandResult commandResult = runCommand("status --short");
+    ImmutableList<String> lines = splitLines(commandResult.stdout);
+    for (String line : lines) {
+      String[] parts;
+      if (!line.trim().startsWith("M")
```
Comment (oferb) [inline common/repo/GitRepo.java]: Why only in these cases do you need to remove a space, and why do you later split by spaces?
Can you document by giving an example of the original line?

File: common/repo/GitRepo.java [hunk 9]
```diff
@@ -169,42 +142,25 @@ public void tagHead(String name) {
...
-                          .build()));
-      status
-          .getUntracked()
-          .forEach(
-              untracked ->
-                  files.add(
-                      File.newBuilder().setAction(File.Action.ADD).setFilename(untracked).build()));
-    } catch (GitAPIException e) {
-      throw new RuntimeException(e);
+    CommandResult commandResult = runCommand("status --short");
```
Comment (oferb) [inline common/repo/GitRepo.java]: Did you check `git status --porcelain`?

File: common/repo/GitRepo.java [hunk 10]
```diff
@@ -366,33 +299,65 @@ public void removeBranch(String branch) {
...
-        }
-        // Index 0 should have file data
-        data = objectReader.open(treeWalk.getObjectId(0)).getBytes();
-      }
-      return new String(data, UTF_8);
-    } catch (Exception e) {
-      throw new RuntimeException(e);
+    CommandResult commandResult = runCommand("show " + commitId + ":" + path);
+    String result = commandResult.stdout;
+    int lastIndexOfNewLineSymbol = result.lastIndexOf("\n");
```
Comment (oferb) [inline common/repo/GitRepo.java]: Can you explain this code? (lines 306-311)

File: common/repo/GitRepo.java [hunk 11]
```diff
@@ -366,33 +299,65 @@ public void removeBranch(String branch) {
...
+  }
+
+  public String getHeadCommitId() {
+    CommandResult commandResult = runCommand("rev-parse HEAD");
+    return commandResult.stdout.replace("\n", "");
+  }
+
+  private void switchToMasterBranch() {
+    if (!currentBranch().equals("master")) {
+      System.out.println("You are not in the branch master");
```
Comment (oferb) [inline common/repo/GitRepo.java]: No need to print these. It's normal not to be in the master branch sometimes, and the calling code can decide to log.

File: common/repo/GitRepo.java [hunk 12]
```diff
@@ -366,33 +299,65 @@ public void removeBranch(String branch) {
...
+      switchBranch("master");
+    }
+  }
+
+  @VisibleForTesting
+  public ImmutableList<String> getTagList() {
+    ImmutableList.Builder<String> result = ImmutableList.builder();
+    CommandResult commandResult = runCommand("tag");
+    try {
+      ImmutableList<String> lines = splitLines(commandResult.stdout);
```
Comment (oferb) [inline common/repo/GitRepo.java]: Why are you looping over the lines, to add them to another list, instead of using that other list (`lines`)?

File: common/repo/GitRepo.java (hunk 9 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: I changed command to `git status --porcelain`.

File: common/repo/GitRepo.java (hunk 7 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: I need `master..`. Changed it.

File: common/repo/GitRepo.java [hunk 13]
```diff
@@ -16,65 +16,36 @@
...
-import org.eclipse.jgit.revwalk.RevWalk;
-import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
-import org.eclipse.jgit.treewalk.TreeWalk;
 
 // TODO: Implement methods
 @AutoFactory
 public class GitRepo implements Repo {
-  private final Git jGit;
-  private final Repository jGitRepo;
+  private static final String TWO_WHITE_SPACES = "\\s{2}";
```
[the remark this answers is not in the document]
Comment (val-fed) [reply common/repo/GitRepo.java]: `git status --porcelain` (or `git status --short`) command shows lines in a different format. Some lines have two spaces between status symbol and file name, others single spaces. For example:
```
git status --porcelain

D  LICENSE
 M common/repo/GitRepo.java
A  common/repo/new_file_ADD.txt
R  README.md -> rename_README.md
?? common/repo/new_file_UNTRACKED.txt

```

File: common/repo/GitRepo.java (hunk 10 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Actually it needed for tests. I some changed tests and deleted this code.

File: common/repo/GitRepo.java [hunk 14]
```diff
@@ -177,18 +149,12 @@ public void tagHead(String name) {
...
     ImmutableList<String> lines = splitLines(commandResult.stdout);
     for (String line : lines) {
-      String[] parts;
-      if (!line.trim().startsWith("M")
-          || !line.trim().startsWith("RM")
-          || !line.trim().startsWith("??")) {
-        line = line.replaceFirst(TWO_WHITE_SPACES, " ");
-      }
-      parts = line.trim().split(" ");
+      line = line.replaceFirst("\\s{2}", " ");
```
Comment (oferb) [inline common/repo/GitRepo.java]: Would `"  "` work instead of `"\\s{2}"`? It's simpler.

File: common/repo/GitRepo.java [hunk 15]
```diff
@@ -177,18 +149,12 @@ public void tagHead(String name) {
...
+      String[] parts = line.trim().split(" ");
 
       File.Builder fileBuilder = File.newBuilder();
```
Comment (oferb) [inline common/repo/GitRepo.java]: I think this is nicer:
```
Action action = getAction(parts[0]);
String filename = action.equals(File.Action.RENAME) ? parts[3] : parts[1]);
files.add(File.newBuilder().setAction(action).setFilename(filename).build());
```

File: common/repo/GitRepo.java [hunk 16]
```diff
@@ -244,107 +211,77 @@ public void tagHead(String name) {
...
   }
 
+  @Override
   public void pushAll() {
-    try {
-      jGit.push().setPushAll().setAtomic(true).call();
-    } catch (GitAPIException e) {
-      throw new RuntimeException(e);
-    }
+    runCommand("push --all origin");
```
Comment (oferb) [inline common/repo/GitRepo.java]: @vmax is this correct?

File: common/repo/GitRepo.java [hunk 17]
```diff
@@ -244,107 +211,77 @@ public void tagHead(String name) {
...
-        add.addFilepattern(file);
-      }
-      try {
-        add.call();
-        jGit.commit().setAll(true).setMessage("Updated changes").call();
-        return true;
-      } catch (NoFilepatternException e) {
-        // Nothing to commit
-        return true;
+    switchToMasterBranch();
```
Comment (oferb) [inline common/repo/GitRepo.java]: @vmax can you review `mergeTheirs()` and `merge()`?

File: common/repo/GitRepo.java [hunk 18]
```diff
@@ -365,23 +302,47 @@ public void removeBranch(String branch) {
...
-    String result = commandResult.stdout;
-    int lastIndexOfNewLineSymbol = result.lastIndexOf("\n");
-    if (lastIndexOfNewLineSymbol >= 0) {
-      result =
-          new StringBuilder(result)
-              .replace(lastIndexOfNewLineSymbol, lastIndexOfNewLineSymbol + 1, "")
-              .toString();
-    }
-    return result;
+    return commandResult.stdout;
```
Comment (oferb) [inline common/repo/GitRepo.java]: You can simplify to:
`return runCommand("show " + commitId + ":" + path).stdout;`

File: common/repo/GitRepo.java [hunk 19]
```diff
@@ -365,23 +302,47 @@ public void removeBranch(String branch) {
...
 
-  public String currentBranch() {
-    return runCommand("rev-parse --abbrev-ref HEAD").stdout.trim();
+  public void addFile(String path) {
+    runCommand("add " + path);
+  }
+
+  public String getHeadCommitId() {
+    CommandResult commandResult = runCommand("rev-parse HEAD");
+    return commandResult.stdout.replace("\n", "");
```
Comment (oferb) [inline common/repo/GitRepo.java]: Maybe also add `trim()`?

File: common/repo/GitRepo.java [hunk 20]
```diff
@@ -365,23 +302,47 @@ public void removeBranch(String branch) {
...
+      switchBranch("master");
+    }
+  }
+
+  @VisibleForTesting
+  public ImmutableList<String> getTagList() {
+    CommandResult commandResult = runCommand("tag");
+    try {
+      return splitLines(commandResult.stdout);
+    } catch (IllegalStateException e) {
```
Comment (oferb) [inline common/repo/GitRepo.java]: The stack trace will include `getTagList()`, so you don't have to explicitly catch & rethrow, plus you're removing the original exception.

File: common/repo/tests/GitRepoTest.java [hunk 21]
```diff
@@ -156,5 +161,149 @@ public void testGetMultipleCommits() {
                 .build()),
         repo.getCommits(TEST_BRANCH));
   }
+
+  @Test
+  public void testAddFile() {
```
Comment (oferb) [inline common/repo/tests/GitRepoTest.java]: Looks like it's not just `add`, so how about:
s/testAddFile/testCommitFile or testAddAndCommitFile`

File: common/repo/tests/GitRepoTest.java [hunk 22]
```diff
@@ -156,5 +161,149 @@ public void testGetMultipleCommits() {
...
+        repo.getUncommittedFiles());
+  }
+
+  @Test
+  public void testGetUncommittedFilesWhenModifiedFile() {
+    repo.switchBranch(TEST_BRANCH);
+    fileUtils.writeStringUnchecked(TEST_FILE_CONTENTS, fileUtils.joinPaths(repoFolder, TEST_FILE));
+    gitRepo.addFile(TEST_FILE);
+    repo.commit(repo.getUncommittedFiles(), COMMIT_MESSAGE);
+    fileUtils.writeStringUnchecked("new file contains", fileUtils.joinPaths(repoFolder, TEST_FILE));
```
Comment (oferb) [inline common/repo/tests/GitRepoTest.java]: You probably meant:
`s/new file contains/new file contents`

File: common/repo/tests/GitRepoTest.java [hunk 23]
```diff
@@ -156,5 +161,149 @@ public void testGetMultipleCommits() {
...
+    repo.commit(repo.getUncommittedFiles(), COMMIT_MESSAGE);
+    fileUtils.deleteFileOrDirectoryIfExistsUnchecked(fileUtils.joinPaths(repoFolder, TEST_FILE));
+    assertEquals(
+        ImmutableList.of(
+            File.newBuilder().setFilename(TEST_FILE).setAction(File.Action.DELETE).build()),
+        repo.getUncommittedFiles());
+  }
+
+  @Test
+  public void testTadHead() {
```
Comment (oferb) [inline common/repo/tests/GitRepoTest.java]: s/testTadHead/testTagHead

File: common/repo/tests/GitRepoTest.java [hunk 24]
```diff
@@ -156,5 +161,149 @@ public void testGetMultipleCommits() {
...
+    repo.tagHead("test_tag");
+    assertEquals(ImmutableList.of("test_tag"), gitRepo.getTagList());
```
Comment (oferb) [inline common/repo/tests/GitRepoTest.java]: Let's also test 2 tags

File: common/repo/GitRepo.java [hunk 25]
```diff
@@ -153,14 +154,11 @@ public void tagHead(String name) {
...
+      line = line.replaceFirst("  ", " ");
       String[] parts = line.trim().split(" ");
-
-      File.Builder fileBuilder = File.newBuilder();
-      fileBuilder.setAction(getAction(parts[0]));
-      fileBuilder.setFilename(
-          fileBuilder.getAction().equals(File.Action.RENAME) ? parts[3] : parts[1]);
-      files.add(fileBuilder.build());
+      Action action = getAction(parts[0]);
+      String filename = action.equals(File.Action.RENAME) ? parts[3] : parts[1];
```
Comment (oferb) [inline common/repo/GitRepo.java]: Can you add an example of what a regular line and a rename line looks like?
```
// We extract the action and filename from line. Here are some example lines:
// <regular line>
// <rename line>
```

File: common/repo/GitRepo.java [hunk 26]
```diff
@@ -244,107 +211,77 @@ public void tagHead(String name) {
...
+              "merge origin/"
+                  + branch
+                  + " -m Merge_remote-tracking_branch_\'origin/"
+                  + branch
+                  + "\'");
+    } else {
+      mergeCommandResult = runCommand("merge " + branch);
     }
+    String err = mergeCommandResult.stderr;
+    return err.length() == 0;
```
Comment (vmax) [inline common/repo/GitRepo.java]: @val-fed `return mergeCommandResult.stderr.length().isEmpty()`

File: common/repo/GitRepo.java [hunk 27]
```diff
@@ -244,107 +211,77 @@ public void tagHead(String name) {
...
+    String err = commandResult.stderr;
+    return err.length() == 0;
+  }
+
+  public boolean merge(String branch, boolean remote) {
+    switchToMasterBranch();
+    CommandResult mergeCommandResult;
+    if (remote) {
+      CommandResult fetchCommandResult = runCommand("fetch origin " + branch);
+      if (fetchCommandResult.stderr.length() != 0) {
```
Comment (vmax) [inline common/repo/GitRepo.java]: @val-fed `if (!fetchCommandResult.stderr.isEmpty())`

File: common/repo/GitRepo.java (hunk 26 above)
Comment (vmax) [inline common/repo/GitRepo.java]: @val-fed @oferb  did you decide to not reset `HEAD` to previous position?

File: common/repo/GitRepo.java (hunk 27 above)
Comment (oferb) [reply common/repo/GitRepo.java]: No we didn't. Good catch!

File: common/repo/repo.proto (hunk 1 above)
Comment (val-fed) [reply common/repo/repo.proto]: Done.

File: common/repo/GitRepo.java (hunk 2 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Done.

File: common/repo/GitRepo.java (hunk 4 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Done.

File: common/repo/GitRepo.java (hunk 6 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Done.

File: common/repo/GitRepo.java (hunk 11 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Done.

File: common/repo/GitRepo.java (hunk 12 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Changed.

File: common/repo/GitRepo.java (hunk 14 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Fixed.

File: common/repo/GitRepo.java (hunk 15 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Fixed.

File: common/repo/GitRepo.java (hunk 16 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Fixed.

File: common/repo/tests/GitRepoTest.java (hunk 21 above)
Comment (val-fed) [reply common/repo/tests/GitRepoTest.java]: Fixed.

File: common/repo/tests/GitRepoTest.java (hunk 22 above)
Comment (val-fed) [reply common/repo/tests/GitRepoTest.java]: Fixed.

File: common/repo/tests/GitRepoTest.java (hunk 23 above)
Comment (val-fed) [reply common/repo/tests/GitRepoTest.java]: Fixed.

File: common/repo/tests/GitRepoTest.java (hunk 24 above)
Comment (val-fed) [reply common/repo/tests/GitRepoTest.java]: Done.

File: common/repo/GitRepo.java (hunk 25 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Done.

File: common/repo/GitRepo.java (hunk 27 above)
Comment (val-fed) [reply common/repo/GitRepo.java]: Fixed.
